Query-frontend will occasionally panic when under sustained load

**Describe the bug**

Query-frontend will occasionally panic:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x47ad05]
goroutine 302 [running]:
errors.(*errorString).Error(0x0, 0x209c7e0, 0x0)
	/usr/local/go/src/errors/errors.go:68 +0x5
github.com/opentracing/opentracing-go/log.Field.Marshal(0x2485b37, 0xc, 0x9, 0x0, 0x0, 0x0, 0x209c7e0, 0x0, 0x2995038, 0xc06a0b00d8)
	/drone/src/vendor/github.com/opentracing/opentracing-go/log/field.go:231 +0x309
github.com/uber/jaeger-client-go.ConvertLogsToJaegerTags(0xc000a68780, 0x3, 0x3, 0xc06984e008, 0x39c7248, 0xc000cfbd78)
	/drone/src/vendor/github.com/uber/jaeger-client-go/jaeger_tag.go:31 +0x13f
github.com/uber/jaeger-client-go.buildLogs(0xc06975e480, 0x1, 0x1, 0x100, 0x39c7248, 0x0)
	/drone/src/vendor/github.com/uber/jaeger-client-go/jaeger_thrift_span.go:82 +0x145
github.com/uber/jaeger-client-go.BuildJaegerThrift(0xc0697586c0, 0x0)
	/drone/src/vendor/github.com/uber/jaeger-client-go/jaeger_thrift_span.go:43 +0x16c
github.com/uber/jaeger-client-go.(*udpSender).Append(0xc000ccad80, 0xc0697586c0, 0x0, 0x0, 0x2)
	/drone/src/vendor/github.com/uber/jaeger-client-go/transport_udp.go:129 +0x45
github.com/uber/jaeger-client-go.(*remoteReporter).processQueue(0xc000533f80)
	/drone/src/vendor/github.com/uber/jaeger-client-go/reporter.go:304 +0x1c2
created by github.com/uber/jaeger-client-go.NewRemoteReporter
	/drone/src/vendor/github.com/uber/jaeger-client-go/reporter.go:237 +0x1a5
```

We observed this while receiving a constant load of about 23 queries per second.

**To Reproduce**
Steps to reproduce the behavior:
1. Run Tempo
2. Send a ton of queries

**Expected behavior**
Tempo should not panic.

**Environment:**
 - Infrastructure: Kubernetes
 - Deployment tool: jsonnet

**Additional Context**

--- a/docs/tempo/website/configuration/_index.md
+++ b/docs/tempo/website/configuration/_index.md
@@ -161,7 +161,7 @@ The Query Frontend is responsible for sharding incoming requests for faster proc
 query_frontend:
 
     # number of times to retry a request sent to a querier
-    # (default: 5)
+    # (default: 2)
     [max_retries: <int>]
 
     # number of shards to split the query into

--- a/modules/frontend/config.go
+++ b/modules/frontend/config.go
@@ -17,7 +17,7 @@ func (cfg *Config) RegisterFlagsAndApplyDefaults(prefix string, f *flag.FlagSet)
 	cfg.Config.DownstreamURL = ""
 	cfg.Config.Handler.LogQueriesLongerThan = 0
 	cfg.Config.FrontendV1.MaxOutstandingPerTenant = 100
-	cfg.MaxRetries = 5
+	cfg.MaxRetries = 2
 	cfg.QueryShards = 2
 }
 

--- a/modules/frontend/frontend.go
+++ b/modules/frontend/frontend.go
@@ -39,7 +39,7 @@ func NewTripperware(cfg Config, logger log.Logger, registerer prometheus.Registe
 		// - the Deduper dedupes Span IDs for Zipkin support
 		// - the ShardingWare shards queries by splitting the block ID space
 		// - the RetryWare retries requests that have failed (error or http status 500)
-		rt := NewRoundTripper(next, Deduper(logger), ShardingWare(cfg.QueryShards, logger), RetryWare(cfg.MaxRetries, logger))
+		rt := NewRoundTripper(next, Deduper(logger), ShardingWare(cfg.QueryShards, logger), RetryWare(cfg.MaxRetries, registerer))
 
 		return queryrange.RoundTripFunc(func(r *http.Request) (*http.Response, error) {
 			start := time.Now()

--- a/modules/frontend/retry.go
+++ b/modules/frontend/retry.go
@@ -52,13 +52,13 @@ func (r retryWare) Do(req *http.Request) (*http.Response, error) {
 		resp, err := r.next.Do(req)
 
 		// do not retry if no error and response is not HTTP 5xx
-		if err == nil && resp.StatusCode/100 != 5 {
+		if err == nil && !shouldRetry(resp.StatusCode) {
 			return resp, nil
 		}
 
 		// do not retry if GRPC error contains response that is not HTTP 5xx
 		httpResp, ok := httpgrpc.HTTPResponseFromError(err)
-		if ok && httpResp.Code/100 != 5 {
+		if ok && !shouldRetry(int(httpResp.Code)) {
 			return resp, err
 		}
 
@@ -87,4 +87,8 @@ func (r retryWare) Do(req *http.Request) (*http.Response, error) {
 			ot_log.String("errMsg", errMsg),
 		)
 	}
+}
+
+func shouldRetry(statusCode int) bool {
+	return statusCode/100 == 5
 }

--- a/modules/frontend/retry.go
+++ b/modules/frontend/retry.go
@@ -3,25 +3,32 @@ package frontend
 import (
 	"net/http"
 
-	"github.com/go-kit/kit/log"
 	"github.com/opentracing/opentracing-go"
 	ot_log "github.com/opentracing/opentracing-go/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/httpgrpc"
 )
 
-func RetryWare(maxRetries int, logger log.Logger) Middleware {
+func RetryWare(maxRetries int, registerer prometheus.Registerer) Middleware {
 	return MiddlewareFunc(func(next Handler) Handler {
 		return retryWare{
 			next:       next,
-			logger:     logger,
 			maxRetries: maxRetries,
+			retriesCount: promauto.With(registerer).NewHistogram(prometheus.HistogramOpts{
+				Namespace: "tempo",
+				Name:      "query_frontend_retries",
+				Help:      "Number of times a request is retried.",
+				Buckets:   []float64{0, 1, 2, 3, 4, 5},
+			}),
 		}
 	})
 }
 
 type retryWare struct {
-	next       Handler
-	logger     log.Logger
-	maxRetries int
+	next         Handler
+	maxRetries   int
+	retriesCount prometheus.Histogram
 }
 
 // Do implements Handler
@@ -33,7 +40,8 @@ func (r retryWare) Do(req *http.Request) (*http.Response, error) {
 	// context propagation
 	req = req.WithContext(ctx)
 
-	triesLeft := r.maxRetries
+	tries := 0
+	defer func() { r.retriesCount.Observe(float64(tries)) }()
 
 	for {
 		if ctx.Err() != nil {
@@ -42,25 +50,23 @@ func (r retryWare) Do(req *http.Request) (*http.Response, error) {
 
 		resp, err := r.next.Do(req)
 
-		triesLeft--
-		if triesLeft == 0 {
-			return resp, err
-		}
-
+		// do not retry if no error and reponse is not HTTP 5xx
 		if err == nil && resp.StatusCode/100 != 5 {
 			return resp, nil
 		}
 
-		statusCode := 0
-		if resp != nil {
-			statusCode = resp.StatusCode
+		// do not retry if GRPC error contains response that is not HTTP 5xx
+		httpResp, ok := httpgrpc.HTTPResponseFromError(err)
+		if ok && httpResp.Code/100 != 5 {
+			return resp, err
 		}
 
-		span.LogFields(
-			ot_log.String("msg", "error processing request"),
-			ot_log.Int("status_code", statusCode),
-			ot_log.Error(err),
-			ot_log.Int("triesLeft", triesLeft),
-		)
+		// reached max retries
+		tries++
+		if tries >= r.maxRetries {
+			return resp, err
+		}
+
+		span.LogFields(ot_log.String("msg", "error processing request. retrying"))
 	}
 }

--- a/modules/frontend/retry.go
+++ b/modules/frontend/retry.go
@@ -50,7 +50,7 @@ func (r retryWare) Do(req *http.Request) (*http.Response, error) {
 
 		resp, err := r.next.Do(req)
 
-		// do not retry if no error and reponse is not HTTP 5xx
+		// do not retry if no error and response is not HTTP 5xx
 		if err == nil && resp.StatusCode/100 != 5 {
 			return resp, nil
 		}

--- a/modules/frontend/retry.go
+++ b/modules/frontend/retry.go
@@ -1,6 +1,7 @@
 package frontend
 
 import (
+	"fmt"
 	"net/http"
 
 	"github.com/opentracing/opentracing-go"
@@ -67,6 +68,23 @@ func (r retryWare) Do(req *http.Request) (*http.Response, error) {
 			return resp, err
 		}
 
-		span.LogFields(ot_log.String("msg", "error processing request. retrying"))
+		statusCode := 0
+		if resp != nil {
+			statusCode = resp.StatusCode
+		}
+		if httpResp != nil {
+			statusCode = int(httpResp.Code)
+		}
+
+		// avoid calling err.Error() on an error returned by frontend tripperware
+		// https://github.com/grafana/tempo/issues/857
+		errMsg := fmt.Sprint(err)
+
+		span.LogFields(
+			ot_log.String("msg", "error processing request. retrying"),
+			ot_log.Int("try", tries),
+			ot_log.Int("status_code", statusCode),
+			ot_log.String("errMsg", errMsg),
+		)
 	}
 }

--- a/modules/frontend/retry_test.go
+++ b/modules/frontend/retry_test.go
@@ -7,8 +7,9 @@ import (
 	"net/http/httptest"
 	"testing"
 
-	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
 	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/httpgrpc"
 	"go.uber.org/atomic"
 )
 
@@ -25,18 +26,20 @@ func TestRetry(t *testing.T) {
 	for _, tc := range []struct {
 		name          string
 		handler       Handler
+		maxRetries    int
 		expectedTries int32
 		expectedRes   *http.Response
 		expectedErr   error
 	}{
 		{
-			name: "retry until success",
+			name: "retry errors until success",
 			handler: HandlerFunc(func(req *http.Request) (*http.Response, error) {
 				if try.Inc() == 5 {
 					return &http.Response{StatusCode: 200}, nil
 				}
 				return nil, errors.New("this request failed")
 			}),
+			maxRetries:    5,
 			expectedTries: 5,
 			expectedRes:   &http.Response{StatusCode: 200},
 			expectedErr:   nil,
@@ -47,18 +50,31 @@ func TestRetry(t *testing.T) {
 				try.Inc()
 				return &http.Response{StatusCode: 400}, nil
 			}),
+			maxRetries:    5,
 			expectedTries: 1,
 			expectedRes:   &http.Response{StatusCode: 400},
 			expectedErr:   nil,
 		},
 		{
+			name: "don't retry GRPC request with HTTP 400's",
+			handler: HandlerFunc(func(req *http.Request) (*http.Response, error) {
+				try.Inc()
+				return nil, httpgrpc.ErrorFromHTTPResponse(&httpgrpc.HTTPResponse{Code: 400})
+			}),
+			maxRetries:    5,
+			expectedTries: 1,
+			expectedRes:   nil,
+			expectedErr:   httpgrpc.ErrorFromHTTPResponse(&httpgrpc.HTTPResponse{Code: 400}),
+		},
+		{
 			name: "retry 500s",
 			handler: HandlerFunc(func(req *http.Request) (*http.Response, error) {
 				try.Inc()
-				return &http.Response{StatusCode: 500}, nil
+				return &http.Response{StatusCode: 503}, nil
 			}),
+			maxRetries:    5,
 			expectedTries: 5,
-			expectedRes:   &http.Response{StatusCode: 500},
+			expectedRes:   &http.Response{StatusCode: 503},
 			expectedErr:   nil,
 		},
 		{
@@ -69,15 +85,38 @@ func TestRetry(t *testing.T) {
 				}
 				return nil, errors.New("not the last request")
 			}),
+			maxRetries:    5,
 			expectedTries: 5,
 			expectedRes:   nil,
 			expectedErr:   errors.New("request failed"),
+		},
+		{
+			name: "maxRetries=1",
+			handler: HandlerFunc(func(req *http.Request) (*http.Response, error) {
+				try.Inc()
+				return &http.Response{StatusCode: 500}, nil
+			}),
+			maxRetries:    1,
+			expectedTries: 1,
+			expectedRes:   &http.Response{StatusCode: 500},
+			expectedErr:   nil,
+		},
+		{
+			name: "maxRetries=0",
+			handler: HandlerFunc(func(req *http.Request) (*http.Response, error) {
+				try.Inc()
+				return &http.Response{StatusCode: 500}, nil
+			}),
+			maxRetries:    0,
+			expectedTries: 1,
+			expectedRes:   &http.Response{StatusCode: 500},
+			expectedErr:   nil,
 		},
 	} {
 		t.Run(tc.name, func(t *testing.T) {
 			try.Store(0)
 
-			retryWare := RetryWare(5, log.NewNopLogger())
+			retryWare := RetryWare(tc.maxRetries, prometheus.NewRegistry())
 			handler := retryWare.Wrap(tc.handler)
 
 			req := httptest.NewRequest("GET", "http://example.com", nil)
@@ -101,7 +140,7 @@ func TestRetry_CancelledRequest(t *testing.T) {
 	req, err := http.NewRequestWithContext(ctx, "GET", "http://example.com", nil)
 	require.NoError(t, err)
 
-	_, err = RetryWare(5, log.NewNopLogger()).
+	_, err = RetryWare(5, prometheus.NewRegistry()).
 		Wrap(HandlerFunc(func(req *http.Request) (*http.Response, error) {
 			try.Inc()
 			return nil, ctx.Err()
@@ -117,7 +156,7 @@ func TestRetry_CancelledRequest(t *testing.T) {
 	req, err = http.NewRequestWithContext(ctx, "GET", "http://example.com", nil)
 	require.NoError(t, err)
 
-	_, err = RetryWare(5, log.NewNopLogger()).
+	_, err = RetryWare(5, prometheus.NewRegistry()).
 		Wrap(HandlerFunc(func(req *http.Request) (*http.Response, error) {
 			try.Inc()
 			cancel()

